core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -118,4 +119,62 @@ class ReplicaManagerTest {
 
     TestUtils.verifyNonDaemonThreadsStatus(this.getClass.getName)
   }
+
+  @Test
+  def testClearPurgatoryOnBecomingFollower() {
+    val props = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect)
+    props.put("log.dir", TestUtils.tempRelativeDir("data").getAbsolutePath)
+    val config = KafkaConfig.fromProps(props)
+    val zkClient = EasyMock.createMock(classOf[ZkClient])
+    val zkUtils = ZkUtils(zkClient, false)
+    val mockLogMgr = TestUtils.createLogManager(config.logDirs.map(new File(_)).toArray)
+    val time = new MockTime()
+    val jTime = new JMockTime
+    val metrics = new Metrics
+    val rm = new ReplicaManager(config, metrics, time, jTime, zkUtils, new MockScheduler(time), mockLogMgr,
+      new AtomicBoolean(false))
+
+    try {
+      var callbackFired = false
+      def callback(responseStatus: Map[TopicPartition, PartitionResponse]) = {
+        assertEquals("Should give NotLeaderForPartitionException", Errors.NOT_LEADER_FOR_PARTITION.code, responseStatus.values.head.errorCode)
+        callbackFired = true
+      }
+
+      val aliveBrokers = Seq(new Broker(0, "host0", 0), new Broker(1, "host1", 1))
+      val metadataCache = EasyMock.createMock(classOf[MetadataCache])
+      EasyMock.expect(metadataCache.getAliveBrokers).andReturn(aliveBrokers).anyTimes()
+      EasyMock.replay(metadataCache)
+
+      val partition = rm.getOrCreatePartition(topic, 0)
+      partition.getOrCreateReplica(0)
+      // Make this replica the leader.
+      val leaderAndIsrRequest1 = new LeaderAndIsrRequest(0, 0,
+        collection.immutable.Map(new TopicPartition(topic, 0) -> new PartitionState(0, 0, 0, Seq(0, 1).asJava.asInstanceOf[java.util.List[Integer]],
+          0, Set(0, 1).asJava.asInstanceOf[java.util.Set[Integer]])).asJava,
+        Set(new BrokerEndPoint(0, "host1", 0), new BrokerEndPoint(1, "host2", 1)).asJava)
+      rm.becomeLeaderOrFollower(0, leaderAndIsrRequest1, metadataCache, (_, _) => {})
+      rm.getLeaderReplicaIfLocal(topic, 0)
+
+      // Append a message.
+      rm.appendMessages(
+        timeout = 1000,
+        requiredAcks = -1,
+        internalTopicsAllowed = false,
+        messagesPerPartition = Map(new TopicPartition(topic, 0) -> new ByteBufferMessageSet(new Message("first message".getBytes))),
+        responseCallback = callback)
+
+      // Make this replica the follower
+      val leaderAndIsrRequest2 = new LeaderAndIsrRequest(0, 0,
+        collection.immutable.Map(new TopicPartition(topic, 0) -> new PartitionState(0, 1, 1, Seq(0, 1).asJava.asInstanceOf[java.util.List[Integer]],
+          0, Set(0, 1).asJava.asInstanceOf[java.util.Set[Integer]])).asJava,
```

Comment:
This is to show that if you pass the right type parameter when constructing the `Seq`, then auto-boxing happens during construction and a lot of ugliness is avoided.